mitmproxy/proxy/layers/dns.py

```diff
@@ -91,24 +113,44 @@ def state_query(self, event: events.Event) -> layer.CommandGenerator[None]:
         from_client = event.connection is self.context.client
 
         if isinstance(event, events.DataReceived):
+            msg = None
             try:
-                msg = dns.Message.unpack(event.data)
+                if self.context.client.transport_protocol == "udp":
+                    msg = dns.Message.unpack(event.data)
+                elif self.context.client.transport_protocol == "tcp":
+                    if self.expected_size is None:
+                        assert self.buf is None
+                        (self.expected_size,) = _LENGTH_LABEL.unpack_from(event.data, 0)
```

Comment:
Nit: this can fail if the client only sends one byte. More realistically, this will fail if the client does query pipelining and the last send frame had only one byte of the next query.